docs/design.md

```diff
@@ -38,13 +38,112 @@ spec:
 ```
 
 ### Command
+
 Optionally you can override the default command of a container by providing
 `command` in the Substratus resource:
 ```yaml
 spec:
   command: ["train.sh"]
 ```
 
+### Storage
+
+#### Bucket Paths
+
+In Kubernetes the combination of the following fields make up a unique reference to a given object (i.e. database key): `split(.apiVersion, "/")[0] + .kind + .metadata.namespace + .metadata.name`. By storing related artifacts using a similar scheme, restore operations are made trivial: the administrator will not need to worry about backing up the `.status.url` field (as opposed to a scheme where storage location is `.metadata.uid` based - which would be unique for a given in-time instance of an object). When deleting and recreating clusters, if the storage bucket persisted, objects (i.e. Models, Datasets) can simply be re-applied into the new cluster (i.e. [velero](https://velero.io/)/similar is not needed). This plays nicely with GitOps.
+
+Storage scheme:
+
+```
+# Models
+gs://{project_id}-substratus-storage/namespaces/{namespace}/models/{name}/model
+gs://{project_id}-substratus-storage/namespaces/{namespace}/models/{name}/logs
+
+# Datasets
+gs://{project_id}-substratus-storage/namespaces/{namespace}/datasets/{name}/data
+gs://{project_id}-substratus-storage/namespaces/{namespace}/datasets/{name}/logs
```

Comment:
I knew S3 had issues when naming paths sequentially but wasn't sure on GCS. This [GCP blog post](https://cloud.google.com/blog/products/gcp/optimizing-your-cloud-storage-performance-google-cloud-performance-atlas) calls out upload performance impact to paths having the same or similar prefixes. Unsure if that applies to downloads.

We could apply a hashing scheme across a string of `"{project_id}-{namespace}-{name}"` as a `{computed_hash}` prefix and likely still be able to query using paths like the following to get all datasets in the bucket for a specific namespace:
```
gs://{project_id}-substratus-storage/**/namespaces/{namespace}/datasets/ *
```

Maybe this is premature optimization but at scale this def matters.

Reply:
Updated PR to including hashing.